jax/random.py

```diff
@@ -417,19 +418,66 @@ def _randint(key, shape, minval, maxval, dtype):
   return lax.add(minval, lax.convert_element_type(random_offset, dtype))
 
 
-def shuffle(key: np.ndarray, x: np.ndarray, axis: int = 0) -> np.ndarray:
-  """Shuffle the elements of an array uniformly at random along an axis.
+def shuffle(key: np.ndarray, x: np.ndarray, axis: Optional[int]=0) -> np.ndarray:
+  """Shuffle the elements of an array uniformly along the first axis.
 
   Args:
     key: a PRNGKey used as the random key.
     x: the array to be shuffled.
-    axis: optional, an int axis along which to shuffle (default 0).
+    axis: (deprecated) an int axis along which to shuffle (default 0). If not None,
+      subarrays are shuffled independently along the specified axis, as in
+      :func:`~jax.random.shuffle_axis`. If `None`, subarrays are shuffled as a group.
+      along the leading axis, as in :func:`numpy.random.shuffle`.
+      In a future release, the default will be changed to `None` and support for
+      integer axis values will be dropped.
 
   Returns:
     A shuffled version of x.
+
+  >>> import jax.numpy as np
+  >>> from jax import random
+  >>> x = np.arange(9).reshape(3, 3)
+  >>> random.shuffle(random.PRGKey(0), x, axis=None)
+  DeviceArray([[3, 4, 5],
+               [6, 7, 8],
+               [0, 1, 2]], dtype=int32)
   """
+  # TODO: drop support for int axis after Jax version 0.1.67
+  if axis is None:
+    return x[_shuffle(key, np.arange(x.shape[0]), 0)]
+  if x.ndim != 1:
+    warnings.warn("shuffle(): int axis with multi-dimenstional input is deprecated. "
+      "Set axis=None to sort groups along the first axis, or use shuffle_independent() "
+      "to sort subarrays independently along a specified axis.", JaxDeprecationWarning)
   return _shuffle(key, x, axis)
 
+
+def shuffle_independent(key: np.ndarray, x: np.ndarray, axis: int=0) -> np.ndarray:
```

Comment:
Should this be `random.permutation`, for consistency with NumPy?

Reply:
Done

Reply:
Oops, wrong comment. I think permutation is different... numpy doesn't have any function to shuffle each subarray independently along an axis. I suppose an alternative would be to drop this functionality entirely.